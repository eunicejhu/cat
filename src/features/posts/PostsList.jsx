import React from "react";
import { useSelector } from "react-redux";
import AddPostForm from "./AddPostForm.tsx";

const PostsList = () => {
  const posts = useSelector((state) => state.posts);
  const renderPosts =
    posts &&
    posts.map((post) => (
      <article className="post-excerpt" key={post.id}>
        <h3>{post.title}</h3>
        <p className="post-content">{post.content.substring(0, 100)}</p>
      </article>
    ));

  return (
    <>
      <section className="add-post-form">
        <AddPostForm />
      </section>
      <section className="posts-list">
        <h2>Posts</h2>
        {posts && posts.length ? renderPosts : "No posts"}
      </section>
    </>
  );
};

export default PostsList;
