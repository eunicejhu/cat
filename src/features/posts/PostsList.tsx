import React from "react";
import { useSelector } from "react-redux";
import { useRouteMatch, Link } from "react-router-dom";
import { isAfter, parseISO } from "date-fns";
import AddPostForm from "./AddPostForm";
import { State } from "../../store";
import PostAuthor from "./PostAuthor";
import TimeAgo from "./TimeAgo";
const PostsList = () => {
  const { path } = useRouteMatch();
  const posts = useSelector((state: State) => state.posts);
  const sortedPosts =
    posts && posts.slice().sort((a, b) => a.date.localeCompare(b.date));
  const renderPosts =
    sortedPosts &&
    sortedPosts.map((post) => (
      <article className="post-excerpt" key={post.id}>
        <h3>{post.title}</h3>
        <p className="post-content">{post.content.substring(0, 100)}</p>
        <PostAuthor userId={post.userId} />
        <TimeAgo date={post.date} />
        <p>
          <Link to={`${path}posts/${post.id}`}>See more</Link>
        </p>
      </article>
    ));

  return (
    <>
      <section className="add-post-form">
        <AddPostForm />
      </section>
      <section className="posts-list">
        <h2>Posts</h2>
        {posts && posts.length ? renderPosts : "No posts"}
      </section>
    </>
  );
};

export default PostsList;
