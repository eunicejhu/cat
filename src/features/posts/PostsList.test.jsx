import React from "react";
import { render } from "@testing-library/react";
import StoreWrapper from "../../test/StoreWrapper.tsx";
import PostsList from "./PostsList";

jest.mock("react-redux", () => {
  return {
    useSelector: jest
      .fn()
      .mockImplementationOnce(jest.requireActual("react-redux").useSelector)
      .mockImplementationOnce(() => [])
      .mockImplementationOnce(() => undefined),
    Provider: jest.requireActual("react-redux").Provider,
  };
});

test("show initial postsList", () => {
  const { container, getByText } = render(<PostsList />, {
    wrapper: StoreWrapper,
  });
  expect(getByText(/First test Post!/i)).toBeInTheDocument();
  expect(container.querySelectorAll(".post-excerpt").length).toBe(2);
});

test("show no posts when posts is []", () => {
  const { getByText } = render(<PostsList />, {
    wrapper: StoreWrapper,
  });
  expect(getByText(/No Posts/i)).toBeInTheDocument();
});
test("show no posts when posts is undefined", () => {
  const { getByText } = render(<PostsList />, {
    wrapper: StoreWrapper,
  });
  expect(getByText(/No Posts/i)).toBeInTheDocument();
});
