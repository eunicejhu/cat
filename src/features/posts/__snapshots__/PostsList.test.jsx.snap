// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render correctly 1`] = `
Array [
  <section
    className="add-post-form"
  >
    <div
      className="add-post"
    >
      <form>
        <div
          className="field"
        >
          <label
            htmlFor="title"
          >
            Title
          </label>
          <input
            data-testid="title"
            id="title"
            name="title"
            onChange={[Function]}
            value=""
            width={20}
          />
        </div>
        <div
          className="field"
        >
          <label
            htmlFor="content"
          >
            Content
          </label>
          <textarea
            data-testid="content"
            id="content"
            name="content"
            onChange={[Function]}
            value=""
          />
        </div>
        <div
          className="field"
        >
          <label
            htmlFor="users"
          >
            Author
          </label>
          <select
            data-testid="users"
            id="users"
            name="users"
            onChange={[Function]}
            value=""
          >
            <option
              value=""
            />
            <option
              value="0"
            >
              Tianna Jenkins
            </option>
            <option
              value="1"
            >
              Kevin Grant
            </option>
            <option
              value="2"
            >
              Madison Price
            </option>
          </select>
        </div>
        <div
          className="field"
        >
          <button
            disabled={true}
            onClick={[Function]}
            type="button"
          >
            Add Post
          </button>
        </div>
      </form>
    </div>
  </section>,
  <section
    className="posts-list"
  >
    <h2>
      Posts
    </h2>
    <article
      className="post-excerpt"
    >
      <h3>
        First test Post!
      </h3>
      <p
        className="post-content"
      >
        test!
      </p>
      <span
        className="post-author"
      >
        By 
        Kevin Grant
      </span>
      <span
        className="date"
      >
         
        <i>
          
        </i>
         
      </span>
      <p>
        <a
          href="/posts/1"
          onClick={[Function]}
        >
          See more
        </a>
      </p>
    </article>
    <article
      className="post-excerpt"
    >
      <h3>
        Second test Post
      </h3>
      <p
        className="post-content"
      >
        test
      </p>
      <span
        className="post-author"
      >
        By 
        Tianna Jenkins
      </span>
      <span
        className="date"
      >
         
        <i>
          
        </i>
         
      </span>
      <p>
        <a
          href="/posts/2"
          onClick={[Function]}
        >
          See more
        </a>
      </p>
    </article>
  </section>,
]
`;
